backend/service/k8s/cache.go

```diff
@@ -0,0 +1,137 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+// TODO (mcutalo): Make this configurable or keep this at a high value.
+// The implications of resync running often is both the pressure it puts on our datastore,
+// and the additional requests that we place on all configured kubernetes cluster.
+// We have to be mindful of our clients Burst & QPS configuration,
+// which is overrideable by the user.
+const informerResyncTime = time.Hour * 1
+
+// Setting a large channel buffer mostly for first boot and the  resync timer,
+// this really should be sized according to the size of your k8s deployment.
+// However this should be a large enough buffer for the datastore to keep up with.
+const topologyObjectChanBufferSize = 5000
```

Comment:
random q, do we have a sense for what kind of memory consumption a value of 5000 will require? how big are k8s objects in the buffer?

Reply:
Good question, I would have to measure the size of our protos. Just to be clear this isn't buffering the entire k8s object just Clutch's v1 k8s protos (pod, hpa, deployment), so there is substantially less data.

Reply:
I've taken the following to calculate rough memory consumption, this example is for a pod.
```
resource := topologyv1.Resource{
	Id: pod.Name,
	Pb: protoPod,
}
update := topologyv1.UpdateCacheRequest{
	Resource: &resource,
	Action:   action,
}
log.Printf("update %d", reflect.TypeOf(update).Size())
log.Printf("resource %d", reflect.TypeOf(resource).Size())
log.Printf("proto %d", reflect.TypeOf(*protoPod).Size())
```
output:
```
2020/10/13 13:35:45 update 56
2020/10/13 13:35:45 resource 72
2020/10/13 13:35:45 proto 80
```

From what I can tell it's hard to measure size recursively, so we looking at the size of the `Resource`, `UpdateCacheRequest` and the protoPod is the output of `ptypes.MarshalAny(pod)`.

(208 payload bytes * 5000) / 1024 / 1024 = 0.99 Mb